internal/testsupport/check.go

```diff
@@ -0,0 +1,96 @@
+package testsupport
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+)
+
+// BackendKey allows to retrieve a resource from the backend using its ID.
+//
+// BackendKey must return true if the resource was found.
+type BackendKey func(c *hcloud.Client, id int) bool
+
+// CheckResourceExists checks that a resource with the passed name exists.
+//
+// CheckResourceExists uses k to actually retrieve the resource from the
+// Hetzner Cloud backend.
+func CheckResourceExists(name string, k BackendKey) resource.TestCheckFunc {
+	const op = "testsupport/CheckResourceExists"
+
+	return func(s *terraform.State) error {
+		if err := backendResourceByKey(s, name, k); err != nil {
+			return fmt.Errorf("%s: %v", op, err)
+		}
+		return nil
+	}
+}
+
+// CheckResourcesDestroyed checks if resources of resType do not exsist in the
+// backend anymore.
+func CheckResourcesDestroyed(resType string, k BackendKey) resource.TestCheckFunc {
+	const op = "testsupport/CheckResourceDestroyed"
+
+	return func(s *terraform.State) error {
+		for name, rs := range s.RootModule().Resources {
+			if rs.Type != resType {
+				continue
+			}
+			err := backendResourceByKey(s, name, k)
+			if err != nil && !errors.Is(err, errMissingInBackend) {
+				return fmt.Errorf("%s: %v", op, err)
+			}
+		}
+		return nil
+	}
+}
+
+var errMissingInBackend = errors.New("missing in backend")
```

Comment:
This error message should be "missing in Hetzner Cloud" or something like that. Backend might be confusing, as this could also be another backend (terraform itself). Using the Term "Hetzner Cloud" could bring the user to the correct solution.